File: src/com/prey/beta/actions/PreyBetaController.java
/*******************************************************************************
 * Created by Carlos Yaconi
 * Copyright 2012 Fork Ltd. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey.beta.actions;

import android.app.NotificationManager;
import android.content.Context;
import android.content.Intent;


import com.prey.PreyConfig;
import com.prey.beta.services.PreyBetaRunnerService;


public class PreyBetaController {

	public static void startPrey(Context ctx) {
		PreyConfig config = PreyConfig.getPreyConfig(ctx);
		if (config.isThisDeviceAlreadyRegisteredWithPrey()){
			// Cancelling the notification of the SMS that started Prey
			NotificationManager nm = (NotificationManager) ctx.getSystemService(Context.NOTIFICATION_SERVICE);
			nm.cancelAll();

			config.setRun(true);
			final Context context = ctx;
			new Thread(new Runnable() {
				
				public void run() {
					//First need to stop a previous running instance.
					context.stopService(new Intent(context, PreyBetaRunnerService.class));
					context.startService(new Intent(context, PreyBetaRunnerService.class));
				}
			}).start();
			
		}
	}

	public static void stopPrey(Context ctx) {
		ctx.stopService(new Intent(ctx, PreyBetaRunnerService.class));
	}

}

File: app/src/main/java/com/prey/beta/actions/PreyBetaController.kt
/*******************************************************************************
 * Created by Orlando Aliaga
 * Copyright 2025 Prey Inc. All rights reserved.
 * License: GPLv3
 * Full license at "/LICENSE"
 ******************************************************************************/
package com.prey.beta.actions

import android.content.Context

import com.prey.PreyConfig
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/**
 * Controller class for Prey beta actions.
 */
class PreyBetaController {

    /**
     * Starts the Prey beta actions with a null command.
     *
     * @param context The application context.
     */
    fun startPrey(context: Context) {
        PreyConfig.getInstance(context).setLastEvent("start_prey")
        startPrey(context, null)
    }

    /**
     * Starts the Prey beta actions with a given command.
     *
     * @param context The application context.
     * @param cmd The command to execute (optional).
     */
    fun startPrey(context: Context, cmd: String?) {
        // Check if the device is already registered with Prey.
        if (PreyConfig.getInstance(context).isThisDeviceAlreadyRegisteredWithPrey()) {
            CoroutineScope(Dispatchers.IO).launch {
                PreyBetaActionsRunner.getInstance(context).execute()
            }
        }
    }

    companion object {
        private var instance: PreyBetaController? = null
        fun getInstance(): PreyBetaController =
            instance ?: PreyBetaController().also { instance = it }
    }

}